### PR TITLE
NATS streaming reconnect handlers for gateway and queue-workers

### PR DESCRIPTION
**Pull request by vosmith**
Signed-off-by: Vincent Smith <vosmith08@gmail.com>

This PR adds support for the NATS streaming clients to reconnect to the NATS Streaming servers when a NATS connection is reset.

## Description

The handler for the gateway and the queue-worker both have wrappers around their NATS connections to handle an `onReconnect` event from the underlying NATS connection.  The gateway re-connects to the NATS Streaming server using the same client and cluster IDs.  The queue-worker re-connects and resubscribes to the `faas-request` topic.

## Motivation and Context

Previously, when a NATS connection was reset, the clients would not be able to communicate with the NATS streaming server. This is due to the NATS streaming server running in in-memory mode [as detailed in this issue](https://github.com/nats-io/nats-streaming-server/issues/150).

- [X] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

openfaas/faas#579

## How Has This Been Tested?

Using the OpenFaaS helm chart, I deployed an OpenFaaS stack on a Kubernetes cluster running Kubernetes 1.8.8.  I deployed a simple echo function using faas-cli.  While tailing the logs of the gateway and the queue-worker pods, I invoked the echo function with --async and watched the logs show the messages being passed successfully.  I killed the NATS container and waited for it to start up again.  I invoked the function with --async again and waited for the publish ack timeout to occur.

Using a modified helm chart and Docker images, I deployed an OpenFaaS stack with the changes from this PR on a the same K8s cluster.  I deployed a simple echo function using faas-cli.  While tailing the logs of the gateway and the queue-worker pods, I invoked the echo function with --async and watched the messages being passed successfully.  I killed the NATS container and waited for it to start up again.   Log message is printed by the queue worker on successful reconnection.  I  invoked the function again and saw the messages being successfully passed.

I also ran the function in a loop, invoking the call every 1 second.  When killing the NATS container, 1 publish ack timeout occurs and the subsequent messages are published successfully.

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [X] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: handler/handler.go [hunk 1]
```diff
@@ -26,7 +27,13 @@ func CreateNatsQueue(address string, port int) (*NatsQueue, error) {
 	clientID := "faas-publisher-" + val
 	clusterID := "faas-cluster"
 
-	nc, err := stan.Connect(clusterID, clientID, stan.NatsURL(natsURL))
+	conn, err := nats.Connect(natsURL, nats.ReconnectHandler(queue1.reconnectClient(clientID, clusterID)))
```
Comment (alexellis) [inline handler/handler.go]: Please can you explain why this changed from `stan` to `nats` and what the implications are?

File: handler/handler.go [hunk 2]
```diff
@@ -47,3 +54,15 @@ func (q *NatsQueue) Queue(req *queue.Request) error {
 
 	return err
 }
+
+func (q *NatsQueue) reconnectClient(clientID string, clusterID string) nats.ConnHandler {
+	return func(c *nats.Conn) {
+		nc, err := stan.Connect(clusterID, clientID, stan.NatsConn(c))
+		if err != nil {
+			log.Println("Failed to reconnect to NATS stream")
```
Comment (alexellis) [inline handler/handler.go]: Does it make sense to bundle up a single error / log statement here with `log.Printf()`?

File: queue-worker/main.go [hunk 3]
```diff
@@ -48,150 +49,163 @@ func makeClient() http.Client {
...
+const (
+	subj      = "faas-request"
+	qgroup    = "faas"
+	clusterID = "faas-cluster"
+)
 
-	clusterID := "faas-cluster"
-	val, _ := os.Hostname()
-	clientID := "faas-worker-" + val
+var (
```
Comment (alexellis) [inline queue-worker/main.go]: I would prefer refactoring for style to be done in subsequent PRs as it blurs the intent of the PR and makes it harder to review the change.

File: queue-worker/main.go [hunk 4]
```diff
@@ -48,150 +49,163 @@ func makeClient() http.Client {
...
 
-		printMsg(msg, i)
+	if err != nil {
+		status = http.StatusServiceUnavailable
 
-		started := time.Now()
+		log.Println(err)
+		timeTaken := time.Since(started).Seconds()
 
-		req := queue.Request{}
```
Comment (alexellis) [inline queue-worker/main.go]: I don't know why this changed

File: queue-worker/main.go [hunk 5]
```diff
@@ -48,150 +49,163 @@ func makeClient() http.Client {
...
+			resultStatusCode, resultErr := postResult(&client, req, functionResult, status)
+			if resultErr != nil {
+				log.Println(resultErr)
+			} else {
+				log.Printf("Posted result: %d", resultStatusCode)
+			}
 		}
 
-		queryString := ""
-		if len(req.QueryString) > 0 {
```
Comment (alexellis) [inline queue-worker/main.go]: Why was this removed?

File: handler/handler.go (hunk 1 above)
Comment (vosmith) [reply handler/handler.go]: In order to to define a `ReconnectHandler`, it must be set directly on the NATS connection.  The NATS streaming connection (created by the `stan` package) accepts an existing NATS connection as a parameter so I created a connection with a `ReconnectHandler` and passed into `stan`.

Alternatively, it should be possible to create a connection through `stan`, then set the `ReconnectHandler` on the underlying NATS Streaming connection with `nc.NatsConn().SetReconnectHandler`

After looking deeper into it, the difference here is we are now responsible for closing the NATS connection.

File: handler/handler.go (hunk 2 above)
Comment (vosmith) [reply handler/handler.go]: Yes that makes more sense.

File: handler/handler.go (hunk 1 above)
Comment (wallyqs) [reply handler/handler.go]: The NATS client has internal reconnection logic already, so I think maybe using the `MaxReconnects` option might fit better here?  When connecting to NATS Streaming via `stan.Connect` internally the STAN client will use the default reconnection logic which is to attempt to connect a few times to the server (reconnecting state) before giving up entirely and closing the connection (closed state which is terminal so client will not reconnect again).
By setting the `ReconnectHandler` and `ClosedHandler` it is possible to add custom logic or logging for after the reconnection or closing of the connection has occurred (for example could make the server exit when closed state is reached meaning that client will not reconnect to NATS again).
It is also possible to make the client reconnect forever rather than giving up by setting `nats.MaxReconnects(-1)` in case that behavior might be better suited too.

Comment (vosmith) [reply handler/handler.go]: Hey @wallyqs.  The underlying NATS connection reconnect as you described. The issue is the NATS Streaming server loses all of it's data because it is in memory mode.  Part of this is the client IDs.  Because of this a new streaming connection has to be made to rejoin the cluster and register itself.

Btw I am working on restructuring this commit to be more reviewer friendly.

Comment (wallyqs) [reply handler/handler.go]: Ok might be missing something then... The reconnect handler is called when the NATS client has successfully established a connection to a new server and the NATS client internally replays all the subscriptions once having reconnected, so for the STAN client the reconnection to NATS would be transparent.

**Comment by alexellis** [comment]
Hey @vosmith perhaps a practical way to move it forward would be to take a backup of the branch where it is now - then start over from master/head and apply the minimum set of changes to make the fix and force push that back up.

After we've merged we can look at the best way to refactor the code.

How does that sound? Do you have an ETA on when this could be ready?

**Comment by vosmith** [comment]
Hey @alexellis . Yes I hope to complete the updates this week. :+1:

File: handler/handler.go (hunk 1 above)
Comment (vosmith) [reply handler/handler.go]: @wallyqs , unfortunately it hasn't worked that way in my experience, and it is documented in https://github.com/nats-io/nats-streaming-server/issues/150 .  I agree with you that the STAN client should resubscribe, and this may might be an issue that could be addressed by that library.  But they don't seem to mind (https://github.com/nats-io/go-nats-streaming/issues/153)

**Comment by vosmith** [comment]
```
The command "curl -sSL test.docker.com | sudo -E sh" failed and exited with 100 during .
```

Hey @alexellis , looks like the build failed due to an issue connecting with the Docker server.  Is there a way to trigger this build again without pushing another commit?

**Comment by kozlovic** [comment]
General comment (will comment specifically on the PR): If you use a NATS Streaming server with memory store, it is true that if the server is restarted, since no state is being restored, the previously "connected" clients will stop receiving messages. Publishers would fail too since the server would reject published messages for unknown client IDs.
The streaming server and streaming clients communicate through some inboxes. When the Streaming server is restarted, since it lost that knowledge, it can't communicate with existing clients. Moreover, even internal subjects used to communicate between the server and its clients contain a unique id that won't be the same after the restart).

Note: If the NATS Streaming server connects to a non-embedded NATS Server, then if the NATS Server itself is restarted, that is fine, the client library's use of the underlying NATS connection will reconnect and everything would work fine (some timeout may occur for the operations that were inflight when the NATS server was restarted). This is because the Streaming server would still be running and its state maintained, so the communication can continue.

File: handler/handler.go [hunk 6]
```diff
@@ -27,6 +28,7 @@ func CreateNatsQueue(address string, port int) (*NatsQueue, error) {
 	clusterID := "faas-cluster"
 
 	nc, err := stan.Connect(clusterID, clientID, stan.NatsURL(natsURL))
+	nc.NatsConn().SetReconnectHandler(queue1.reconnectClient(clientID, clusterID, natsURL))
```
Comment (kozlovic) [inline handler/handler.go]: As it is, you should check for err because if there was an error, `nc` would be `nil` and you would get a panic here.

But this is a moot point. You should be creating the NATS connection first, configured with the reconnect handler. Then pass the NATS connection to the `stan.Connect()` call. Otherwise, you still potentially miss a window between the `stan.Connect()` call and the `nc.NatsConn().SetReconnectHandler()` where the connection may have disconnect/reconnect. It will also be easier on reconnect to simply recreate the stan connection (no need to close the connection for which you just got a notification that you were reconnected).

You would need to store the NATS connection (and close it after closing the stan connection) in your NatsQueue structure. By the way, since you would need to store both, I would recommend `sc` for stan connection and `nc` for nats connection. Right now, `nc` is used for the stan connection.

File: handler/handler.go [hunk 7]
```diff
@@ -47,3 +49,16 @@ func (q *NatsQueue) Queue(req *queue.Request) error {
 
 	return err
 }
+
+func (q *NatsQueue) reconnectClient(clientID, clusterID, natsURL string) nats.ConnHandler {
+	return func(c *nats.Conn) {
+		c.Close()
+		nc, err := stan.Connect(clusterID, clientID, stan.NatsURL(natsURL))
```
Comment (kozlovic) [inline handler/handler.go]: See above. If you create the NATS connection, you should here close the stan connection (not the nats one), just for cleanup, then call stan.Connect() again and pass the `c` connection with the `stan.NatsConn()` option. (you would not need the `NatsURL()` option btw).

File: queue-worker/main.go [hunk 8]
```diff
@@ -210,9 +211,28 @@ func main() {
...
+		log.Println("Wait for ", ackWait)
+		sub, err = sc.QueueSubscribe(subj, qgroup, mcb, startOpt, stan.DurableName(durable), stan.MaxInflight(maxInflight), stan.AckWait(ackWait))
+		if err != nil {
+			log.Panicln(err)
+		}
+
+		log.Printf("Listening on [%s], clientID=[%s], qgroup=[%s] durable=[%s]\n", subj, clientID, qgroup, durable)
+	}
+
+	sc.NatsConn().SetReconnectHandler(reconnectHandler)
```
Comment (kozlovic) [inline queue-worker/main.go]: Same story than describe in `handler.go`

**Comment by kozlovic** [comment]
Unrelated, but in queue-worker, both `durable` and `unsubscribe` are defined but not used at all. Unless you have plans to add durable and unsubscribe support, I would have remove that to keep the code cleaner.

**Comment by vosmith** [comment]
@kozlovic, thanks for the feedback!  I'll have those changes ready soon :+1:

**Comment by alexellis** [comment]
@vosmith try to keep the changes (refactoring/style) to an absolute minimum.

File: handler/handler.go [hunk 9]
```diff
@@ -16,6 +16,8 @@ type NatsQueue struct {
 	nc stan.Conn
 }
 
+var natsConn *nats.Conn
```
Comment (kozlovic) [inline handler/handler.go]: You don't need a global var...

File: handler/handler.go [hunk 10]
```diff
@@ -27,8 +29,15 @@ func CreateNatsQueue(address string, port int) (*NatsQueue, error) {
...
+	natsConn, err = nats.Connect(natsURL, nats.ReconnectHandler(queue1.reconnectClient(clientID, clusterID, natsURL)))
+	if err != nil {
+		return nil, err
+	}
+
+	nc, err := stan.Connect(clusterID, clientID, stan.NatsConn(natsConn))
+	if err != nil {
+		return nil, err
+	}
 	queue1.nc = nc
```
Comment (kozlovic) [inline handler/handler.go]: You should store in queue1 also the nats connection in order to close (if the code is ever closing on exit, etc..)

File: handler/handler.go [hunk 11]
```diff
@@ -52,13 +61,13 @@ func (q *NatsQueue) Queue(req *queue.Request) error {
 
 func (q *NatsQueue) reconnectClient(clientID, clusterID, natsURL string) nats.ConnHandler {
 	return func(c *nats.Conn) {
-		c.Close()
-		nc, err := stan.Connect(clusterID, clientID, stan.NatsURL(natsURL))
+		q.nc.Close()
+		c.SetReconnectHandler(q.reconnectClient(clientID, clusterID, natsURL))
```
Comment (kozlovic) [inline handler/handler.go]: Do not reset the reconnect handler. The nats connection is still valid.

File: queue-worker/main.go [hunk 12]
```diff
@@ -28,6 +28,8 @@ type AsyncReport struct {
 	TimeTaken    float64 `json:"timeTaken"`
 }
 
+var natsConn *nats.Conn
```
Comment (kozlovic) [inline queue-worker/main.go]: Don't need this here

File: queue-worker/main.go [hunk 13]
```diff
@@ -80,11 +82,17 @@ func main() {
 	var durable string
 	var qgroup string
 	var unsubscribe bool
+	var err error
 
 	client := makeClient()
-	sc, err := stan.Connect(clusterID, clientID, stan.NatsURL("nats://"+natsAddress+":4222"))
+	natsConn, err = nats.Connect("nats://" + natsAddress + ":4222")
```
Comment (kozlovic) [inline queue-worker/main.go]: Set the reconnect handler as an option of `nats.Connect()`

File: queue-worker/main.go [hunk 14]
```diff
@@ -230,7 +240,7 @@ func main() {
 		log.Printf("Listening on [%s], clientID=[%s], qgroup=[%s] durable=[%s]\n", subj, clientID, qgroup, durable)
 	}
 
-	sc.NatsConn().SetReconnectHandler(reconnectHandler)
+	natsConn.SetReconnectHandler(reconnectHandler)
```
Comment (kozlovic) [inline queue-worker/main.go]: Remove this

File: queue-worker/main.go [hunk 15]
```diff
@@ -215,12 +223,14 @@ func main() {
 	var sub stan.Subscription
 
 	reconnectHandler = func(nc *nats.Conn) {
-		nc.Close()
-		sc, err = stan.Connect(clusterID, clientID, stan.NatsURL("nats://"+natsAddress+":4222"))
+		sc.Close()
+		nc.SetReconnectHandler(reconnectHandler)
```
Comment (kozlovic) [inline queue-worker/main.go]: Remove these 2 lines (set reconnect and natsConn=nc). The reconnect handler's `nc *nats.Conn` will be the pointer to the connection you created earlier. No need to do anything with that.

File: handler/handler.go (hunk 10 above)
Comment (kozlovic) [reply handler/handler.go]: Note: I don't see anywhere a Close() or the like in the handler, but still, keep track of it in case this is changed later.

**Comment by vosmith** [comment]
Thanks @kozlovic.  Your input here has been invaluable!

File: queue-worker/main.go [hunk 16]
```diff
@@ -210,9 +208,37 @@ func main() {
 			ackWait = ackWaitVal
 		}
 	}
+	var reconnectHandler nats.ConnHandler
+	var sub stan.Subscription
+	var sc stan.Conn
+	reconnectHandler = func(nc *nats.Conn) {
+		sc.Close()
+
+		sc, err = stan.Connect(clusterID, clientID, stan.NatsConn(nc))
```
Comment (kozlovic) [inline queue-worker/main.go]: From here on, it repeats with what you do below for the initial setup. I would have made the `stan.Connect()` and creating of the queue sub a function (anonymous or not). You may, or may not need to have the print statements in that function, not sure what @alexellis preference would be. In the reconnect handler, you could however log that there was a reconnection.

**Comment by alexellis** [comment]
@vosmith what are your thoughts.
?

**Comment by vosmith** [comment]
Hey @alexellis , after my first attempt I was trying to keep the changes to a minimum in this PR.  I'd like to do a more comprehensive of this file if you don't mind.

**Comment by alexellis** [comment]
Reviewed with contributors on 4/19. 

Hi @vosmith have you been able to address all the feedback from @kozlovic? I would like to see that done and merged before doing a refactoring exercise.

Thanks,

Alex

**Comment by vosmith** [comment]
@alexellis, yes sir.  I've implemented all the changes except for the last suggestion about DRYing up the code.  That is the part I would like to take care of in the refactor.

All the previous suggestions have been Implemented

File: queue-worker/main.go (hunk 16 above)
Comment (alexellis) [reply queue-worker/main.go]: I think we should definitely log this. cc @vosmith

File: handler/handler.go (hunk 10 above)
Comment (alexellis) [reply handler/handler.go]: @vosmith have you addressed this?

**Comment by alexellis** [comment]
I am not seeing the push on this branch since the comments by @kozlovic

File: queue-worker/main.go (hunk 16 above)
Comment (alexellis) [reply queue-worker/main.go]: @vosmith this is just debug information - not DRY - could we log any reconnects?

File: handler/handler.go (hunk 10 above)
Comment (vosmith) [reply handler/handler.go]: Hey @alexellis , there is no place to close this as the connection should stay open as long as the webserver is up and running.  AFAIK there isn't a way to handle a graceful shutdown.

File: queue-worker/main.go (hunk 16 above)
Comment (vosmith) [reply queue-worker/main.go]: Hey @alexellis , I can definitly log the reconnect here.  I apologize, I was focussed more on the first part of the comment about making the connection a separate function.  I forgot about the comment on logging.

**Comment by ivanayov** [comment]
@vosmith seems like your PR has conflicts?
What's left in order to merge?

**Comment by vosmith** [comment]
@ivanayov, work on #33 has superceded this.  Sorry I should have closed this earlier!

**Comment by ivanayov** [comment]
Derek close
